keps/sig-testing/0014-20180523-test-cluster-framework.md

```diff
@@ -0,0 +1,169 @@
+-->
+
+## Summary
+
+> Making testing with clusters consistent
+
+We propose an abstraction so that users and tools (e.g. CI) have a consistent
+way to spin up and configure clusters they can test against. This abstraction
+is targeted mainly to the needs of the following personas and should work
+equally well used interactively and non-interactively / automated.
+
+Alice, an experienced CLI Dev (see [Key Persona Alice][alice]), finds that she can’t
+easily onboard new contributors like Bob (see [Key Persona Bob][bob]), in part
+because her CLI and tests are too tightly coupled with heavyweight
+infrastructure in `k8s.io/kubernetes`. To solve this problem, Alice sponsors
+the development of a new lightweight testing infrastructure outside
+`k8s.io/kubernetes`.
+
+Meanwhile, Carol, an API/Controller dev (see [Key Persona Carol][carol]), is
+frustrated that there are already too many ways to spin up test clusters.
+
+There is a clear tension between Alice and Carol’s frustrations. Left to their
+own devices, Alice is likely to contribute to the proliferation of test
+infrastructure that frustrates Carol, and Carol may find herself advocating
+that Alice be prevented from acting to resolve her own frustration with the
+status quo.
+
+Most of the content of this KEP has already [been worked on and fleshed
+out][the_doc] mostly by people from sig-testing, more specifically people
+working on the [testing-commons subproject][testing_commons_notes].
+
+[key_personas]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.czu8o1uor9qk
+[alice]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.17sl3t5a8nu
+[bob]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.mecveaw12b4d
+[carol]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.4i5imqvwxv3a
+[testing_commons_notes]: https://docs.google.com/document/d/1TOC8vnmlkWw6HRNHoe5xSv5-qv7LelX6XK3UVCHuwb0/edit#
+[the_doc]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#
+
+## Motivation
+
+With this proposal we would like to address these three points of friction:
+- For Carol, we introduce an abstraction which is intended to unify existing
+  and future methods of spinning up test clusters.
+- For Alice’s frustration, we build a lightweight testing framework with no
+  dependency on `k8s.io/kubernetes`
+- For the tension between Alice and Carol, we ensure that Alice’s framework is
+  fully accessible through Carol’s abstraction.
+
+
+<!--
+TODO: Add note that we did interviews and talked to different people, noted in
+      the "Validation" section of the personas?
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+1. To design an abstraction of “standing up a kubernetes cluster” which is:
```

Comment:
How does this differ from what kubetest is trying to accomplish?

Reply:
We didn't know about `kubetest` before. The intent seems indeed very similar! From an (admittedly very quick) first try there might be room for some UX improvements, specifically for the use case of (new-ish) contributors that want to run it as part of their unit/integration test suite (and would be after fast feedback and low setup overhead).

We'll keep looking at `kubetest` and come back with more specific suggestions.

Is there additional documentation besides the README in the [`test-infra` repo](https://github.com/kubernetes/test-infra/tree/master/kubetest) and the godocs?

cc @mariantalla

Reply:
Yeah `kubetest` is hard to use on its own right now, but I'd be interested in trying to fix that. 

The other documentation is: https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md (`hack/e2e.go` is just a wrapper to call `kubetest`)

Reply:
We're trying to standardize the client tooling abstraction for re-use.  I think there is some overlap of concerns, but there are other constraints we are trying to solve here.